index.bs

```diff
@@ -5952,6 +5961,144 @@ However, [=authenticators=] that do not utilize [[!FIDO-CTAP]] do not necessaril
+        ;       over (aaguid || dpk).
+        ;   (2) the attestation certificate or public key, and supporting certificates,
+        ;       if any.
+        ;
+        ; Note that there are details dependent upon the particular attestation
+        ; statement format.
+        ; See https://www.w3.org/TR/webauthn/#sctn-defined-attestation-formats.
+
+        $$attStmtType,
+    }
+
+    ```
+
+: Authenticator extension processing
+:: For both [=authenticatorMakeCredential=] and [=authenticatorGetAssertion=] operations:
+    1. Create or select the [=user credential=] as usual, and let `userCredentialId` be its <code>[=credentialId=]</code>.
+
+    1. If a [=hardware-bound device key pair=] does not already exist for this [=user credential=] on the [=authenticator=], create it using the same public key algorithm as that used by the [=user credential=]'s [=credential key pair=], otherwise locate the existing [=device-bound key=].
+
+    1. Let `dpk` be the newly created or existing [=device public key=], in COSE_Key format [=credentialPublicKey|in the same fashion as for the user credential's credentialPublicKey=] when the latter is conveyed in [=attested credential data=].
+
+    1. Let `devicePrivateKey` be the newly created or existing [=device private key=].
+
+    1. Let `clientDataHash` be the [=hash of the serialized client data=].
+
+    1. Let `sig` be the result of signing the concatenation of `clientDataHash` and `userCredentialId` using the `devicePrivateKey` and the signature algorithm appropriate for the `devicePrivateKey`'s public key algorithm: `sign((clientDataHash || userCredentialId), devicePrivateKey)`.
+
+    1. Let `scope` have the value zero (0x00) if this is an "entire-device" [=device public key=]. Otherwise, let `scope` have the value one (0x01), indicating a more narrow per-app scope.
+
+    1. Let the `$$attStmtType` values be the result of generating an attestation statement in the [=attestation statement format=] appropriate for the [=user credential=], although substituting the authenticator's <code>[=aaguid=]</code> for |authenticatorData|, and substituting `dpk` for |clientDataHash| in the [=attestation statement format=]'s signing procedure (see [[#sctn-generating-an-attestation-object]]). Attestation statement formats define the `fmt` and `attStmt` members of the `$$attStmtType` expansion.
+
+        In summary, the `$$attStmtType` values typically contain a signature value calculated over the bytes of `(aaguid || dpk)`, the attestation certificate or public key, and supporting certificates, if any.
+
+        Note: The details of the `$$attStmtType` values are dependent upon the particular [=attestation statement=] format. See [[#sctn-attestation-formats]].
+
+    1. Let the `devicePubKey` [=authenticator extension output=] be a [=CBOR=] map as defined by `attObjForDevicePublicKey` above, whose values for its `sig`, `aaguid`, `dpk`, `scope`, and `$$attStmtType` members are as calculated in the prior steps.
```

Comment:
I think some of the variables introduced here should use the `|variable|` syntax rather than just `fixed-width` formatting, no? Especially ones like `devicePrivateKey` that are "internal" variables that don't appear in the output. But this also conflates some such "internal" variables with the corresponding outputs, and to be precise I think we should separate those two. Maybe something like this:

```suggestion
    1. Create or select the [=user credential=] as usual, and let |userCredentialId| be its <code>[=credentialId=]</code>.

    1. If a [=hardware-bound device key pair=] does not already exist for this [=user credential=] on the [=authenticator=], create it using the same public key algorithm as that used by the [=user credential=]'s [=credential key pair=], otherwise locate the existing [=device-bound key=].

    1. Let |dpk| be the newly created or existing [=device public key=], in COSE_Key format [=credentialPublicKey|in the same fashion as for the user credential's credentialPublicKey=] when the latter is conveyed in [=attested credential data=].

    1. Let |aaguid| be the [=authenticator=]'s [=AAGUID=].

    1. Let |devicePrivateKey| be the newly created or existing [=device private key=].

    1. Let |clientDataHash| be the [=hash of the serialized client data=].

    1. Let the `devicePubKey` [=authenticator extension output=] be a [=CBOR=] map as defined by `attObjForDevicePublicKey` above, with members as follows:

        1. Let the `aaguid` member be |aaguid|.

        1. Let the `dpk` member be |dpk|.

        1. Let the `sig` member be the result of signing the concatenation of |clientDataHash| and |userCredentialId| using the |devicePrivateKey| and the signature algorithm appropriate for the |devicePrivateKey|'s public key algorithm: <code>sign((|clientDataHash| || |userCredentialId|), |devicePrivateKey|)</code>.

        1. Let the `scope` member have the value zero (0x00) if this is an "entire-device" [=device public key=]. Otherwise, let `scope` have the value one (0x01), indicating a more narrow per-app scope.

        1. Let the `$$attStmtType` values be the result of generating an attestation statement in the [=attestation statement format=] appropriate for the [=user credential=], although substituting |aaguid| for |authenticatorData|, and substituting |dpk| for |clientDataHash| in the [=attestation statement format=]'s signing procedure (see [[#sctn-generating-an-attestation-object]]). Attestation statement formats define the `fmt` and `attStmt` members of the `$$attStmtType` expansion.

            In summary, the `$$attStmtType` values typically contain a signature value calculated over the bytes of <code>(|aaguid| || |dpk|)</code>, the attestation certificate or public key, and supporting certificates, if any.

            Note: The details of the `$$attStmtType` values are dependent upon the particular [=attestation statement=] format. See [[#sctn-attestation-formats]].

```

Reply:
Thanks @emlun, I overall agree with (and like) this reformulation, although there's details that I suggest polishing to follow prior practice elsewhere in the spec ("authnr extension output _value_"), to use CBOR terminology, and to fix/polish  inconsistencies/blemishes (some of which are holdovers from the original)):
```
s/ `devicePubKey` [=authenticator extension output=] be /  `devicePubKey` [=authenticator extension output=] value be /

s/ with members as follows / with keys and values as follows /

s/ member be / key's value be /g  (ie 3 occurrences)

s/ entire-device / entire device /   (ie cleaning up a prior mistake of mine)

s/ `scope` member / `scope` key /

s/ Let the `$$attStmtType` values / Let the values of the `$$attStmtType` "group socket" [=CDDL=] /   (just a thought...)

s/ appropriate for the [=user credential=] / appropriate for this authenticator /   (see [1] below)

s/ substituting |aaguid| for |authenticatorData| / substituting |aaguid| for `authenticatorData` /

s/ substituting |dpk| for |clientDataHash| substituting |dpk| for `clientDataHash` /

s/ the `$$attStmtType` values typically contain / the `$$attStmtType` values generated by the foregoing procedure typically contain /
```
In terms of process, you or I could edit your above uber-suggestion with those of the above suggestions of mine we agree on, then commit the result...?

[1]  in https://github.com/w3c/webauthn/pull/1663#discussion_r743994007 @agl outlines rationale wrt the DPK's attestation being dependent upon the authenticator's characteristics.

Reply:
Those edits all look good to me! Would you like to make the combined update?

Reply:
Ok, thanks, here's what I've done:  I merged your suggestions as-is in d52342c, and then polished them in e23c4b9 and b8ec5b8.

Reply:
Nice, looks good!